src/interp/interp-inl.h

```diff
@@ -139,95 +139,107 @@ inline Frame::Frame(Ref func,
       mod(mod) {}
 
 //// FreeList ////
-template <typename T>
-bool FreeList<T>::IsUsed(Index index) const {
-  return index < list_.size() && !is_free_[index];
+template <>
+inline bool FreeList<Ref>::IsUsed(Index index) const {
+  return (list_[index].index & refFreeBit) == 0;
 }
 
-template <typename T>
+template <>
+inline FreeList<Ref>::~FreeList<Ref>() {}
+
+template <>
 template <typename... Args>
-auto FreeList<T>::New(Args&&... args) -> Index {
-  if (!free_.empty()) {
-    Index index = free_.back();
-    assert(is_free_[index]);
-    free_.pop_back();
-    is_free_[index] = false;
-    list_[index] = T(std::forward<Args>(args)...);
-    return index;
+auto FreeList<Ref>::New(Args&&... args) -> Index {
+  if (free_head_ == 0) {
+    list_.push_back(Ref(std::forward<Args>(args)...));
+    return list_.size() - 1;
   }
-  assert(list_.size() == is_free_.size());
-  is_free_.push_back(false);
-  list_.emplace_back(std::forward<Args>(args)...);
-  return list_.size() - 1;
-}
 
-template <typename T>
-void FreeList<T>::Delete(Index index) {
-  list_[index] = T();
-  is_free_[index] = true;
-  free_.push_back(index);
-}
+  Index index = free_head_ - 1;
 
-template <typename T>
-const T& FreeList<T>::Get(Index index) const {
-  assert(IsUsed(index));
-  return list_[index];
+  assert(!IsUsed(index));
+  assert(free_items_ > 0);
+
+  free_head_ = list_[index].index & (refFreeBit - 1);
+  list_[index] = Ref(std::forward<Args>(args)...);
+  free_items_--;
+  return index;
 }
 
-template <typename T>
-T& FreeList<T>::Get(Index index) {
+template <>
+inline void FreeList<Ref>::Delete(Index index) {
   assert(IsUsed(index));
-  return list_[index];
+
+  list_[index].index = free_head_ | refFreeBit;
+  free_head_ = index + 1;
+  free_items_++;
 }
 
 template <typename T>
-auto FreeList<T>::size() const -> Index {
-  return list_.size();
+bool FreeList<T>::IsUsed(Index index) const {
+  return (reinterpret_cast<uintptr_t>(list_[index]) & ptrFreeBit) == 0;
 }
 
 template <typename T>
-auto FreeList<T>::count() const -> Index {
-  return list_.size() - free_.size();
+FreeList<T>::~FreeList<T>() {
+  for (auto object : list_) {
+    if ((reinterpret_cast<uintptr_t>(object) & ptrFreeBit) == 0) {
+      delete object;
+    }
+  }
```

Comment:
Where these just leaked before?

Reply:
It was a `std::unique_ptr` before, but casting a struct into a number is not necessary working as expected (implementation dependent).

Reply:
Can we switch back to `std::unique_ptr` and use `.get()` to grap its underlying pointer?